File: packages/wasm-sdk/zig/fluxbase.zig
//! fluxbase — Fluxbase WASM SDK for Zig
//!
//! Import this module in your handler:
//!
//! ```zig
//! const fluxbase = @import("fluxbase");
//!
//! pub const handle = fluxbase.makeHandle(myHandler);
//! pub const __flux_alloc = fluxbase.__flux_alloc;
//!
//! fn myHandler(ctx: *fluxbase.Ctx, payload: []const u8, alloc: std.mem.Allocator) ![]const u8 {
//!     ctx.log("my-function: executing");
//!     return std.fmt.allocPrint(alloc, "{{\"output\":{{\"message\":\"Hello!\"}}}}", .{});
//! }
//! ```
//!
//! ## Build
//! ```
//! zig build -Dtarget=wasm32-wasip1 -Doptimize=ReleaseSmall
//! ```

const std = @import("std");

// ── Host imports ──────────────────────────────────────────────────────────────

extern "fluxbase" fn log(level: i32, ptr: [*]const u8, len: i32) void;
extern "fluxbase" fn secrets_get(
    key_ptr: [*]const u8, key_len: i32,
    out_ptr: [*]u8, out_max: i32,
) i32;
extern "fluxbase" fn http_fetch(
    req_ptr: [*]const u8, req_len: i32,
    out_ptr: [*]u8, out_max: i32,
) i32;

// ── Allocator ─────────────────────────────────────────────────────────────────

var gpa = std.heap.GeneralPurposeAllocator(.{}){};
pub const allocator = gpa.allocator();

/// ABI-required allocator export.
pub export fn __flux_alloc(size: i32) i32 {
    const buf = allocator.alloc(u8, @intCast(size)) catch return 0;
    return @intCast(@intFromPtr(buf.ptr));
}

// ── Ctx ───────────────────────────────────────────────────────────────────────

pub const Ctx = struct {
    /// Emit an info-level log message.
    pub fn logInfo(self: *Ctx, msg: []const u8) void {
        _ = self;
        log(1, msg.ptr, @intCast(msg.len));
    }

    /// Emit a log message at an explicit level (1=info, 2=warn, 3=error).
    pub fn logLevel(self: *Ctx, level: i32, msg: []const u8) void {
        _ = self;
        log(level, msg.ptr, @intCast(msg.len));
    }

    /// Retrieve a secret. Returns null if not found.
    pub fn secret(self: *Ctx, key: []const u8) ?[]const u8 {
        _ = self;
        var buf: [4096]u8 = undefined;
        const n = secrets_get(key.ptr, @intCast(key.len), &buf, buf.len);
        if (n < 0) return null;
        const result = allocator.alloc(u8, @intCast(n)) catch return null;
        @memcpy(result, buf[0..@intCast(n)]);
        return result;
    }

    /// Perform an outbound HTTP request.
    /// `request_json` must match the Fluxbase HTTP request schema.
    /// Returns the response JSON or null on error.
    pub fn fetch(self: *Ctx, request_json: []const u8) ?[]const u8 {
        _ = self;
        const out = allocator.alloc(u8, 65536) catch return null;
        const n = http_fetch(
            request_json.ptr, @intCast(request_json.len),
            out.ptr, @intCast(out.len),
        );
        if (n < 0) return null;
        const result = allocator.alloc(u8, @intCast(n)) catch return null;
        @memcpy(result, out[0..@intCast(n)]);
        return result;
    }
};

// ── Result helpers ────────────────────────────────────────────────────────────

/// Write a result JSON string into `[4-byte LE len][json]` layout and return
/// the pointer to the buffer.  The buffer is heap-allocated and must NOT be
/// freed while the host is reading it.
pub fn writeResult(json: []const u8) i32 {
    const len: u32 = @intCast(json.len);
    const buf = allocator.alloc(u8, 4 + json.len) catch return 0;
    buf[0] = @intCast(len & 0xff);
    buf[1] = @intCast((len >> 8) & 0xff);
    buf[2] = @intCast((len >> 16) & 0xff);
    buf[3] = @intCast((len >> 24) & 0xff);
    @memcpy(buf[4..], json);
    return @intCast(@intFromPtr(buf.ptr));
}

/// Convenience: write `{"output": <outputJson>}`.
pub fn writeOutput(output_json: []const u8) i32 {
    const json = std.fmt.allocPrint(
        allocator,
        "{{\"output\":{s}}}",
        .{output_json},
    ) catch return 0;
    return writeResult(json);
}

/// Convenience: write `{"error": "<message>"}`.
pub fn writeError(message: []const u8) i32 {
    const json = std.fmt.allocPrint(
        allocator,
        "{{\"error\":\"{s}\"}}",
        .{message},
    ) catch return 0;
    return writeResult(json);
}

// ── Handler factory ───────────────────────────────────────────────────────────

/// The signature expected by `makeHandle`.
pub const HandlerFn = fn (ctx: *Ctx, payload: []const u8, alloc: std.mem.Allocator) anyerror![]const u8;

/// Generate a `handle` export function that wraps your handler.
///
/// Usage:
/// ```zig
/// pub const handle = fluxbase.makeHandle(myHandlerFn);
/// ```
pub fn makeHandle(comptime handler: HandlerFn) fn (i32, i32) callconv(.C) i32 {
    return struct {
        pub export fn handle(payload_ptr: i32, payload_len: i32) i32 {
            const payload: []const u8 = @as([*]const u8, @ptrFromInt(@as(usize, @intCast(payload_ptr))))[0..@intCast(payload_len)];
            var ctx = Ctx{};
            const result_json = handler(&ctx, payload, allocator) catch |err| {
                const msg = @errorName(err);
                const json = std.fmt.allocPrint(
                    allocator,
                    "{{\"error\":\"{s}\"}}",
                    .{msg},
                ) catch return 0;
                return writeResult(json);
            };
            const json = std.fmt.allocPrint(
                allocator,
                "{{\"output\":{s}}}",
                .{result_json},
            ) catch return 0;
            return writeResult(json);
        }
    }.handle;
}

File: packages/wasm-sdk/go/fluxbase.go
// Package fluxbase is the Fluxbase WASM SDK for Go (TinyGo).
//
// Build your function with:
//
//	tinygo build -o handler.wasm -target wasip1 -scheduler none -no-debug .
//
// # ABI contract
//
// The Fluxbase runtime calls:
//
//	handle(payload_ptr i32, payload_len i32) i32
//
// The returned i32 is a pointer into linear memory where:
//
//	[4-byte u32 LE length][UTF-8 JSON]
//
// JSON must contain either {"output": ...} or {"error": "..."}.
//
// # Usage
//
//	func main() {
//	    fluxbase.Register(func(ctx *fluxbase.Ctx, payload []byte) (any, error) {
//	        var inp MyInput
//	        json.Unmarshal(payload, &inp)
//	        return MyOutput{Message: "Hello " + inp.Name}, nil
//	    })
//	}
package fluxbase

import (
	"encoding/json"
	"unsafe"
)

// ── Host imports ──────────────────────────────────────────────────────────────

//go:wasmimport fluxbase log
func hostLog(level int32, ptr unsafe.Pointer, length int32)

//go:wasmimport fluxbase secrets_get
func hostSecretsGet(keyPtr unsafe.Pointer, keyLen int32, outPtr unsafe.Pointer, outMax int32) int32

//go:wasmimport fluxbase http_fetch
func hostHttpFetch(reqPtr unsafe.Pointer, reqLen int32, outPtr unsafe.Pointer, outMax int32) int32

// ── Ctx ───────────────────────────────────────────────────────────────────────

// Ctx provides access to Fluxbase host capabilities within a handler.
type Ctx struct{}

// Log emits a log line at the given level (1=info, 2=warn, 3=error).
func (c *Ctx) Log(msg string) {
	b := []byte(msg)
	if len(b) == 0 {
		return
	}
	hostLog(1, unsafe.Pointer(&b[0]), int32(len(b)))
}

// LogLevel emits a log line at a specific level.
func (c *Ctx) LogLevel(level int, msg string) {
	b := []byte(msg)
	if len(b) == 0 {
		return
	}
	hostLog(int32(level), unsafe.Pointer(&b[0]), int32(len(b)))
}

// Secret retrieves a secret by key. Returns ("", false) if not found.
func (c *Ctx) Secret(key string) (string, bool) {
	kb := []byte(key)
	buf := make([]byte, 4096)
	n := hostSecretsGet(
		unsafe.Pointer(&kb[0]), int32(len(kb)),
		unsafe.Pointer(&buf[0]), int32(len(buf)),
	)
	if n < 0 {
		return "", false
	}
	return string(buf[:n]), true
}

// HttpRequest is the input to Ctx.Fetch.
type HttpRequest struct {
	Method  string            `json:"method"`
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
	// Body should be base64-encoded when non-empty.
	Body string `json:"body,omitempty"`
}

// HttpResponse is the result of Ctx.Fetch.
type HttpResponse struct {
	Status  int               `json:"status"`
	Headers map[string]string `json:"headers,omitempty"`
	// Body is base64-encoded.
	Body string `json:"body"`
}

// Fetch performs an outbound HTTP request via the Fluxbase host.
func (c *Ctx) Fetch(req HttpRequest) (*HttpResponse, error) {
	reqBytes, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	out := make([]byte, 65536)
	n := hostHttpFetch(
		unsafe.Pointer(&reqBytes[0]), int32(len(reqBytes)),
		unsafe.Pointer(&out[0]), int32(len(out)),
	)
	if n < 0 {
		return nil, &FetchError{code: int(n)}
	}
	var resp HttpResponse
	if err := json.Unmarshal(out[:n], &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// FetchError is returned when a host http_fetch call fails.
type FetchError struct{ code int }

func (e *FetchError) Error() string {
	return "http_fetch failed with code " + itoa(e.code)
}

func itoa(i int) string {
	if i == 0 {
		return "0"
	}
	neg := i < 0
	if neg {
		i = -i
	}
	buf := [20]byte{}
	pos := len(buf)
	for i > 0 {
		pos--
		buf[pos] = byte('0' + i%10)
		i /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}

// ── Handler registration ──────────────────────────────────────────────────────

// HandlerFunc is the signature of a Fluxbase WASM handler.
type HandlerFunc func(ctx *Ctx, payload []byte) (any, error)

var registeredHandler HandlerFunc

// Register sets the function that will handle incoming invocations.
// Call it from main().
func Register(fn HandlerFunc) {
	registeredHandler = fn
}

// ── WASM ABI exports ──────────────────────────────────────────────────────────

// resultBuf holds the most-recent handler result so the GC doesn't collect it
// before the host reads it.
var resultBuf []byte

//export __flux_alloc
func fluxAlloc(size int32) int32 {
	buf := make([]byte, size)
	return int32(uintptr(unsafe.Pointer(&buf[0])))
}

//export handle
func handle(payloadPtr int32, payloadLen int32) int32 {
	payload := unsafe.Slice((*byte)(unsafe.Pointer(uintptr(payloadPtr))), payloadLen)

	ctx := &Ctx{}
	out, err := registeredHandler(ctx, payload)

	var resultJSON []byte
	if err != nil {
		type errResp struct {
			Error string `json:"error"`
		}
		resultJSON, _ = json.Marshal(errResp{Error: err.Error()})
	} else {
		type okResp struct {
			Output any `json:"output"`
		}
		resultJSON, _ = json.Marshal(okResp{Output: out})
	}

	// Prepend 4-byte LE length
	length := uint32(len(resultJSON))
	resultBuf = make([]byte, 4+len(resultJSON))
	resultBuf[0] = byte(length)
	resultBuf[1] = byte(length >> 8)
	resultBuf[2] = byte(length >> 16)
	resultBuf[3] = byte(length >> 24)
	copy(resultBuf[4:], resultJSON)
	return int32(uintptr(unsafe.Pointer(&resultBuf[0])))
}
